commcare_connect/opportunity/helpers.py

```diff
@@ -357,85 +357,148 @@ def get_payment_report_data(opportunity: Opportunity, usd=False):
-        base_filter |= Q(managedopportunity__program__organization=organization)
-
-    queryset = Opportunity.objects.filter(base_filter).annotate(
-        program=F("managedopportunity__program__name"),
-        pending_invites=pending_invites_subquery(),
-        pending_approvals=Coalesce(pending_approvals_sq, Value(0)),
-        total_accrued=total_accrued_sq(),
-        total_paid=total_paid_sq(),
-        payments_due=ExpressionWrapper(
-            F("total_accrued") - F("total_paid"),
-            output_field=DecimalField(),
-        ),
-        inactive_workers=inactive_workers_subquery(three_days_ago),
-        status=Case(
-            When(Q(active=True) & Q(end_date__gte=today), then=Value(0)),  # Active
-            When(Q(active=True) & Q(end_date__lt=today), then=Value(1)),  # Ended
-            default=Value(2),  # Inactive
-            output_field=IntegerField(),
-        ),
-    )
-
-    if program_manager:
-        total_deliveries_sq = Subquery(
-            CompletedWork.objects.filter(opportunity_access__opportunity_id=OuterRef("pk"))
-            .values("opportunity_access__opportunity_id")
-            .annotate(total=Sum("saved_completed_count"))
-            .values("total")[:1],
-            output_field=IntegerField(),
+class TieredQueryset:
+    # This queryset enables targetted count, order_by, pagination
+    #   for large querysets on which it may not be efficient to run
+    #   them.
+    #
+    # This works by passing a base_queryset that just returns ids
+    #   and data_qs_fn which returns the data for those ids
+    def __init__(self, base_qs, data_qs_fn):
+        self.base_qs = base_qs
+        self.data_qs_fn = data_qs_fn
+        self._ordered_qs = base_qs  # default unless order_by is called
+
+    def count(self):
+        return self.base_qs.count()
+
+    def __getitem__(self, key):
+        if isinstance(key, slice):
+            page_qs = self._ordered_qs[key]
+            ids = list(page_qs.values_list("id", flat=True))
+            return list(self.data_qs_fn(ids))
+        else:
+            obj = self._ordered_qs[key]
+            result = list(self.data_qs_fn([obj.id]))
+            if not result:
+                raise IndexError("Index out of range.")
+            return result[0]
+
+    def __iter__(self):
+        return iter(self[:])
+
+    def order_by(self, *fields):
+        self._ordered_qs = self.base_qs.order_by(*fields)
+        return self
+
+
+class OpportunityData:
+    def __init__(self, org, is_program_manager):
+        self.org = org
+        self.is_program_manager = is_program_manager
+
+    def get_data(self):
+        base_qs = self.get_opportunity_list_id_qs(self.org, self.is_program_manager)
+
+        def data_qs(ids):
+            return self.get_opportunity_list_data(ids, self.is_program_manager)
+
+        return TieredQueryset(base_qs, data_qs)
+
+    @staticmethod
+    def get_opportunity_list_id_qs(organization, program_manager=False):
```

Comment:
Is this method name right? From the name I get the impression that it's returning ids (or a queryset with id's), but the actual return is an Opportunity queryset it seems.

Reply:
Also interested in this question. The description mentions only returning IDs, but this seems like a normal queryset, plus some extra annotations

Reply:
This indeed is used to get just the IDs [above](https://github.com/dimagi/commcare-connect/pull/692/files#diff-396037c378e64671cea6adc7185fa4f3aee28145c1fcf11f27bd597e494cd9a5R378), though I could see how that's not clear from the name. I will rename it to `get_base_qs` which is more apt.